src/os/bluestore/io_uring.cc

```diff
@@ -0,0 +1,394 @@
+  }
+
+  d->mmap[1].ptr = d->sqes;
+
+  d->mmap[2].len = p->cq_off.cqes +
+    p->cq_entries * sizeof(struct io_uring_cqe);
+  ptr = mmap(0, d->mmap[2].len, PROT_READ | PROT_WRITE,
+	     MAP_SHARED | MAP_POPULATE, d->ring_fd,
+	     IORING_OFF_CQ_RING);
+  if (ptr == MAP_FAILED) {
+    munmap(d->mmap[0].ptr, d->mmap[0].len);
+    munmap(d->mmap[1].ptr, d->mmap[1].len);
+    return -errno;
+  }
+
+  d->mmap[2].ptr = ptr;
+  cring->head = (unsigned int *)((char *)ptr + p->cq_off.head);
+  cring->tail = (unsigned int *)((char *)ptr + p->cq_off.tail);
+  cring->ring_mask = (unsigned int *)((char *)ptr + p->cq_off.ring_mask);
+  cring->ring_entries = (unsigned int *)((char *)ptr + p->cq_off.ring_entries);
+  cring->cqes = (struct io_uring_cqe *)((char *)ptr + p->cq_off.cqes);
+  d->cq_ring_mask = *cring->ring_mask;
+
+  return 0;
+}
+
+static void ioring_unmap(struct ioring_data *d)
+{
+  unsigned int i;
+
+  for (i = 0; i < sizeof(d->mmap)/sizeof(d->mmap[0]); i++)
+    munmap(d->mmap[i].ptr, d->mmap[i].len);
+  close(d->ring_fd);
+}
+
+static int ioring_cqring_reap(struct ioring_data *d, unsigned int events,
+                              unsigned int max, struct aio_t **paio)
+{
+  struct io_cq_ring *ring = &d->cq_ring;
+  struct io_uring_cqe *cqe;
+  struct aio_t *io;
+
+  unsigned head, reaped = 0;
+
+  head = *ring->head;
+  do {
+    read_barrier();
+    if (head == *ring->tail)
+      break;
+    cqe = &ring->cqes[head & d->cq_ring_mask];
+    if (cqe->flags & IOCQE_FLAG_CACHEHIT)
+      d->cachehit++;
+    else
+      d->cachemiss++;
+
+    io = (struct aio_t *)(uintptr_t) cqe->user_data;
+    io->rval = cqe->res;
+
+    paio[reaped + events] = io;
+
+    reaped++;
+    head++;
+  } while (reaped + events < max);
+
+  d->queued -= reaped;
+  *ring->head = head;
+  write_barrier();
+
+  return reaped;
+}
+
+static int ioring_getevents(struct ioring_data *d, unsigned int min,
+                            unsigned int max, struct aio_t **paio)
+{
+  unsigned actual_min = iodepth_batch_complete_min == 0 ? 0 : min;
+  struct io_cq_ring *ring = &d->cq_ring;
+  unsigned events = 0;
+  int r;
+
+  if (!d->queued)
+    return 0;
+
+  d->cq_ring_off = *ring->head;
+  do {
+    r = ioring_cqring_reap(d, events, max, paio);
+    if (r) {
+      events += r;
+      continue;
+    }
+
+    if (!sqpoll_thread) {
+      r = io_uring_enter(d, 0, actual_min,
+			 IORING_ENTER_GETEVENTS);
+      if (r < 0) {
+	if (errno == EAGAIN)
+	  continue;
+	printf("io_uring_enter: errno %d\n", errno);
+	break;
+      }
+    }
+  } while (events < min);
+
+  return r < 0 ? r : events;
+}
+
+static int find_fixed_fd(struct ioring_data *d, int real_fd)
+{
+  auto it = d->fixed_fds_map.find(real_fd);
+  if (it == d->fixed_fds_map.end())
+    return -1;
+
+  return it->second;
+}
+
+static void init_io(struct ioring_data *d, unsigned index, struct aio_t *io)
+{
+  struct io_uring_sqe *sqe = &d->sqes[index];
+
+  if (io->iocb.aio_lio_opcode == IO_CMD_PWRITEV)
```

Comment:
Why don't you want to print and crash in all the cases which are not readv/writev? You found places where something else can be issued?

Reply:
Mostly I was just trying to clean up the non-readv/writev handling using ceph_abort_msg instead of assert, but it looks like we should be able to handle noop cleanly at least.  POLL apparently was never implemented but you are right that we should probably abort there rather than nop.  FDSYNC I'm aborting on and fsync passing through.  I think if POLL is changed to abort, the only question then is fsync?

Reply:
Since nobody sends other than write/read I would always print & abort. If someone introduces other requests - io_uring engine has to be updated as well. Seems this is more explicit behavior which forces developer to test if something new is introduced.